shared/typetracking/codeql/typetracking/TypeTracking.qll

```diff
@@ -130,499 +301,510 @@ private module Cached {
       or
       exists(ContentFilter filter | result = tbt |
         step = WithContent(filter) and
-        content = filter.getAMatchingContent()
+        content.asSome() = filter.getAMatchingContent()
         or
         step = WithoutContent(filter) and
-        not content = filter.getAMatchingContent()
+        not content.asSome() = filter.getAMatchingContent()
       )
     )
     or
-    exists(TypeTrackerContent loadContents, boolean hasReturn |
-      exists(TypeTrackerContent storeContents |
+    exists(Content loadContents, boolean hasReturn |
+      exists(Content storeContents |
         step = StoreStep(pragma[only_bind_into](storeContents)) and
-        tbt = MkTypeBackTracker(hasReturn, loadContents) and
+        tbt = contentTypeBackTracker(hasReturn, loadContents) and
         compatibleContents(storeContents, loadContents) and
         result = noContentTypeBackTracker(hasReturn)
       )
       or
       step = LoadStep(pragma[only_bind_into](loadContents)) and
       tbt = noContentTypeBackTracker(hasReturn) and
-      result = MkTypeBackTracker(hasReturn, loadContents)
+      result = contentTypeBackTracker(hasReturn, loadContents)
     )
     or
-    exists(
-      TypeTrackerContent currentContent, TypeTrackerContent store, TypeTrackerContent load,
-      boolean hasCall
-    |
+    exists(Content currentContent, Content store, Content load, boolean hasCall |
       step = LoadStoreStep(pragma[only_bind_into](load), pragma[only_bind_into](store)) and
       compatibleContents(store, pragma[only_bind_into](currentContent)) and
-      tbt = MkTypeBackTracker(pragma[only_bind_into](hasCall), currentContent) and
-      result = MkTypeBackTracker(pragma[only_bind_out](hasCall), load)
+      tbt = contentTypeBackTracker(pragma[only_bind_into](hasCall), currentContent) and
+      result = contentTypeBackTracker(pragma[only_bind_out](hasCall), load)
     )
   }
 
   /**
-   * Gets the summary that corresponds to having taken a forwards
-   * heap and/or intra-procedural step from `nodeFrom` to `nodeTo`.
-   *
-   * Steps contained in this predicate should _not_ depend on the call graph.
+   * Holds if there is flow from `localSource` to `dst` using zero or more
+   * `simpleLocalSmallStep`s.
    */
-  cached
-  predicate stepNoCall(TypeTrackingNode nodeFrom, TypeTrackingNode nodeTo, StepSummary summary) {
-    exists(Node mid | nodeFrom.flowsTo(mid) and smallstepNoCall(mid, nodeTo, summary))
+  pragma[nomagic]
+  predicate flowsTo(LocalSourceNode localSource, Node dst) {
+    dst = localSource
```

Comment:
Perhaps change the type of `localSource` to `Node` and move the type check here, as in Ruby/JS, to avoid a redundant type check in the recursive case (if it is still relevant).

Reply:
It is relevant. I've added a minimal fix and verified that it works. See also https://github.com/github/codeql-core/issues/1044